lib/kubernetes-deploy/kubernetes_resource/custom_resource_definition.rb

```diff
@@ -2,8 +2,86 @@
 module KubernetesDeploy
   class CustomResourceDefinition < KubernetesResource
     TIMEOUT = 2.minutes
+    ROLLOUT_CONFIG_ANNOTATION = "kubernetes-deploy.shopify.io/monitor-instance-rollout"
     GLOBAL = true
 
+    class RolloutConfig
+      attr_reader :success_queries, :failure_queries
+
+      def initialize(rollout_config)
+        config_json = JSON.parse(rollout_config)
+        config = {
+          success_queries: config_json["success_queries"] || default_success_query,
+          failure_queries: config_json["failure_queries"] || default_failure_query,
+        }.deep_symbolize_keys
+
+        # Preemptively create JsonPath objects
+        config[:success_queries].map! do |query|
+          query.update(query) { |k, v| k == :path ? JsonPath.new(v) : v }
+        end
+        config[:failure_queries].map! do |query|
+          query.update(query) { |k, v| k == :path || k == :error_msg_path ? JsonPath.new(v) : v }
+        end
+
+        validate_config(config)
```

Comment:
Generally speaking I've found it is a best practice to keep initializers lean and limit their ability to raise exceptions, so I'm a bit concerned about the fact that this one runs validations. What do you think about requiring the JSON to be parsed before this object is created (e.g. the signature might be `RolloutConfig.new(success_queries: [], failure_queries: [])`) and then having an explicit `valid?` method on it? We could call that from `validate_definition` on both CRDs and CR instances. That would also force us to think about other possible invalid states, e.g. if you have failure conditions but no success conditions or vice-versa, is that ok?

Reply:
I like that idea and had the same reservations about raising inside the constructor